File: aead.py
"""
AEAD shell around the chaos PWLCM core — the SIMPLE, SAFE interface.

This is the "deep module, simple interface" layer. The chaotic keystream in engine.py is
the untouched heart; this file wraps it so a caller cannot trip the three foot-guns that
adversarial testing exposed:

  1. Weak keys        -> handled: keys go through a hash KDF (engine.from_master), and the
                         weak-parameter band is rejected. You cannot pick a bad key.
  2. Keystream reuse  -> handled: every seal() draws a FRESH RANDOM nonce, so the same
                         message encrypted twice yields different output. No two-time pad.
  3. Tampering        -> handled: encrypt-then-MAC (HMAC-SHA256). open() verifies the tag
                         in constant time BEFORE decrypting and raises if anything changed.
  4. Key-confusion    -> handled: a key-COMMITMENT (#6) binds the blob to exactly one key, so a
                         single ciphertext cannot be made to open under two different keys (the
                         attack that breaks AES-GCM / ChaCha20-Poly1305). See commit.py.

Interface is just two calls:

    blob = seal(master_key, plaintext, aad=b"")
    plaintext = open_(master_key, blob, aad=b"")     # raises InvalidTag on tamper/wrong key

Wire format (bytes):  nonce(16) || commit(32) || ciphertext(N) || tag(32)

STILL UNVETTED. This makes the engine structurally CORRECT (it now looks like a real AEAD,
same shape as ChaCha20-Poly1305). It does NOT make the underlying chaos math proven-secure.
See REPORT.md.
"""

from __future__ import annotations

import hashlib
import hmac
import os

from commit import COMMIT_LEN, key_commitment, verify_commitment
from constants import DEFAULT_N_MAPS
from multimap import MultiMapEngine

NONCE_LEN = 16
TAG_LEN = 32           # HMAC-SHA256
_MAC_INFO = b"chaos-pwlcm-v1|mac-key"


class InvalidTag(Exception):
    """Raised when authentication fails — wrong key, or the ciphertext was tampered with."""


def _mac_key(master_key: bytes) -> bytes:
    """Derive an independent MAC key from the master key (domain-separated from the
    keystream, which is derived with a different label inside the engine)."""
    return hmac.new(master_key, _MAC_INFO, hashlib.sha256).digest()


def _tag(master_key: bytes, nonce: bytes, commit: bytes, aad: bytes, ciphertext: bytes) -> bytes:
    """Authenticate nonce + commitment + AAD + ciphertext (encrypt-then-MAC). Length-prefix AAD so
    (aad, ct) boundaries can't be shifted by an attacker. Covering the commitment means the whole
    blob sits under one integrity boundary."""
    m = hmac.new(_mac_key(master_key), digestmod=hashlib.sha256)
    m.update(nonce)
    m.update(commit)
    m.update(len(aad).to_bytes(8, "big"))
    m.update(aad)
    m.update(ciphertext)
    return m.digest()


def seal(master_key: bytes, plaintext: bytes, aad: bytes = b"",
         n_maps: int = DEFAULT_N_MAPS, nonce: bytes | None = None) -> bytes:
    """Encrypt + authenticate. Returns nonce || ciphertext || tag.

    Keystream comes from `n_maps` independent chaotic maps XOR-combined (default 4 — the
    multi-body design that defeats the single-map state-recovery attack). A fresh random nonce
    is generated every call, so encrypting the same plaintext twice gives different output and
    keystream reuse cannot happen.

    `nonce` is optional and exists ONLY to pin the one source of nondeterminism for a known-answer
    test / Rust-parity vector (mirrors streaming.seal_stream's `salt=`). Leave it None in production
    — a fresh random nonce is the safe default. Callers that pass it (e.g. ratchet_aead's session
    KAT path) are responsible for never reusing one under the same key."""
    if not isinstance(master_key, (bytes, bytearray)):
        raise TypeError("master_key must be bytes")
    if nonce is None:
        nonce = os.urandom(NONCE_LEN)
    elif len(nonce) != NONCE_LEN:
        raise ValueError(f"nonce must be exactly {NONCE_LEN} bytes")
    ciphertext = MultiMapEngine(master_key, nonce, n_maps).encrypt(plaintext)
    commit = key_commitment(master_key, nonce, aad)
    tag = _tag(master_key, nonce, commit, aad, ciphertext)
    return nonce + commit + ciphertext + tag


def open_(master_key: bytes, blob: bytes, aad: bytes = b"",
          n_maps: int = DEFAULT_N_MAPS) -> bytes:
    """Verify + decrypt. Raises InvalidTag if the key is wrong or anything was tampered
    with — the plaintext is NEVER returned for a bad tag. `n_maps` must match the value used
    by seal()."""
    if len(blob) < NONCE_LEN + COMMIT_LEN + TAG_LEN:
        raise InvalidTag("ciphertext too short / malformed")
    nonce = blob[:NONCE_LEN]
    commit = blob[NONCE_LEN:NONCE_LEN + COMMIT_LEN]
    tag = blob[-TAG_LEN:]
    ciphertext = blob[NONCE_LEN + COMMIT_LEN:-TAG_LEN]

    expected = _tag(master_key, nonce, commit, aad, ciphertext)
    if not hmac.compare_digest(expected, tag):       # constant-time: no timing leak
        raise InvalidTag("authentication failed — wrong key or tampered ciphertext")
    # Key-commitment (#6): this blob must commit to exactly THIS key. Independent of the tag, so the
    # property holds even if the MAC-key derivation had a weakness. Constant-time compare inside.
    if not verify_commitment(master_key, nonce, aad, commit):
        raise InvalidTag("key-commitment failed — blob not committed to this key")

    return MultiMapEngine(master_key, nonce, n_maps).decrypt(ciphertext)


if __name__ == "__main__":
    key = b"my shared secret key (any bytes)"
    msg = b"Attack at dawn. Wire $40,000 by Friday."

    blob = seal(key, msg)
    print(f"sealed ({len(blob)} bytes): {blob.hex()[:48]}...")
    print(f"opened: {open_(key, blob)!r}")

    # same message twice -> different blobs (fresh nonce, no two-time pad)
    print(f"two seals differ: {seal(key, msg) != seal(key, msg)}")

    # tamper one byte in the ciphertext -> rejected
    bad = bytearray(blob)
    bad[NONCE_LEN + COMMIT_LEN] ^= 0x01
    try:
        open_(key, bytes(bad))
        print("TAMPER NOT DETECTED  <-- BUG")
    except InvalidTag as e:
        print(f"tamper rejected: {e}")

    # wrong key -> rejected
    try:
        open_(b"the wrong key................", blob)
        print("WRONG KEY ACCEPTED  <-- BUG")
    except InvalidTag as e:
        print(f"wrong key rejected: {e}")

File: rust/src/aead.rs
//! Committing AEAD shell — the SIMPLE, SAFE interface over the chaos keystream.
//!
//! Port of aead.py + commit.py (Phase 8.1). The chaos keystream is the (UNVETTED)
//! bulk cipher; integrity and key-commitment ride vetted HMAC-SHA256.
//!
//! Wire format: nonce(16) || commit(32) || ciphertext(N) || tag(32)
//!
//! STILL UNVETTED. See REPORT.md.

use crate::utils::{ct_eq, hmac_sha256, hmac_sha256_multi};
use crate::multimap::MultiMapEngine;

pub const NONCE_LEN: usize = 16;
pub const COMMIT_LEN: usize = 32;
pub const TAG_LEN: usize = 32;

const COMMIT_KEY_INFO: &[u8] = b"chaos-pwlcm-v1|commit-key|v1";
const MAC_INFO: &[u8] = b"chaos-pwlcm-v1|mac-key";

/// CMT-4 key-commitment binding the master key to (salt, aad). Mirror of commit.py.
pub fn key_commitment(master_key: &[u8], salt: &[u8], aad: &[u8]) -> [u8; 32] {
    let k_c = hmac_sha256(master_key, COMMIT_KEY_INFO);
    let alen = (aad.len() as u64).to_be_bytes();
    hmac_sha256_multi(&k_c, &[salt, &alen, aad])
}

/// Encrypt-then-MAC tag over nonce + commitment + length-prefixed aad + ciphertext.
fn aead_tag(master_key: &[u8], nonce: &[u8], commit: &[u8], aad: &[u8], ct: &[u8]) -> [u8; 32] {
    let mac_key = hmac_sha256(master_key, MAC_INFO);
    let alen = (aad.len() as u64).to_be_bytes();
    hmac_sha256_multi(&mac_key, &[nonce, commit, &alen, aad, ct])
}

/// Seal: nonce || commit || (plaintext XOR keystream) || tag. Deterministic in the given nonce.
pub fn aead_seal(
    master_key: &[u8],
    nonce: &[u8],
    plaintext: &[u8],
    aad: &[u8],
    n_maps: usize,
) -> Vec<u8> {
    let ks = MultiMapEngine::new(master_key, nonce, n_maps).keystream(plaintext.len());
    let ct: Vec<u8> = plaintext.iter().zip(ks.iter()).map(|(p, k)| p ^ k).collect();
    let commit = key_commitment(master_key, nonce, aad);
    let tag = aead_tag(master_key, nonce, &commit, aad, &ct);
    let mut blob = Vec::with_capacity(NONCE_LEN + COMMIT_LEN + ct.len() + TAG_LEN);
    blob.extend_from_slice(&nonce[..NONCE_LEN]);
    blob.extend_from_slice(&commit);
    blob.extend_from_slice(&ct);
    blob.extend_from_slice(&tag);
    blob
}

/// Open: verify tag (constant-time) THEN the key-commitment, then decrypt. None on any failure.
pub fn aead_open(
    master_key: &[u8],
    blob: &[u8],
    aad: &[u8],
    n_maps: usize,
) -> Option<Vec<u8>> {
    if blob.len() < NONCE_LEN + COMMIT_LEN + TAG_LEN {
        return None;
    }
    let nonce = &blob[..NONCE_LEN];
    let commit = &blob[NONCE_LEN..NONCE_LEN + COMMIT_LEN];
    let tag = &blob[blob.len() - TAG_LEN..];
    let ct = &blob[NONCE_LEN + COMMIT_LEN..blob.len() - TAG_LEN];

    let expected_tag = aead_tag(master_key, nonce, commit, aad, ct);
    if !ct_eq(&expected_tag, tag) {
        return None;
    }
    let expected_commit = key_commitment(master_key, nonce, aad);
    if !ct_eq(&expected_commit, commit) {
        return None;
    }
    let ks = MultiMapEngine::new(master_key, nonce, n_maps).keystream(ct.len());
    Some(ct.iter().zip(ks.iter()).map(|(c, k)| c ^ k).collect())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::DEFAULT_N_MAPS;

    #[test]
    fn aead_seal_open_roundtrip() {
        let key = b"a shared secret key";
        let nonce = b"sixteen-byte-non";
        let aad = b"context";
        let pt = b"the quick brown fox jumps over the lazy dog";
        let blob = aead_seal(key, nonce, pt, aad, DEFAULT_N_MAPS);
        assert_eq!(aead_open(key, &blob, aad, DEFAULT_N_MAPS).as_deref(), Some(&pt[..]));
    }

    #[test]
    fn aead_open_rejects_tamper_wrong_key_and_aad() {
        let key = b"a shared secret key";
        let nonce = b"sixteen-byte-non";
        let aad = b"context";
        let pt = b"secret payload";
        let blob = aead_seal(key, nonce, pt, aad, DEFAULT_N_MAPS);
        let mut bad = blob.clone();
        let i = NONCE_LEN + COMMIT_LEN + 1;
        bad[i] ^= 0x01;
        assert!(aead_open(key, &bad, aad, DEFAULT_N_MAPS).is_none());
        assert!(aead_open(b"the wrong key", &blob, aad, DEFAULT_N_MAPS).is_none());
        assert!(aead_open(key, &blob, b"other", DEFAULT_N_MAPS).is_none());
        assert!(aead_open(key, &blob[..blob.len() - 1], aad, DEFAULT_N_MAPS).is_none());
    }
}
